File: frontend/src/services/sites.ts
import { apiClient } from "./base";

export interface PublishedSite {
  slug: string;
  version: number;
  url: string;
  visibility: string;
  sizeBytes: number;
  expiresAt: string | null;
  conversationId: string;
  unpublishedAt: string | null;
  entry: string;
  fileCount?: number | null;
}

export interface PublishSitePayload {
  conversationId: string;
  source?: string;
  slug?: string;
  visibility?: "unlisted" | "public";
}

export const sitesAPI = {
  listMine: async (): Promise<PublishedSite[]> => {
    return await apiClient.get("/sites/me");
  },
  publish: async (payload: PublishSitePayload): Promise<PublishedSite> => {
    const body: Record<string, unknown> = {
      conversationId: payload.conversationId,
      visibility: payload.visibility || "unlisted",
    };
    if (payload.source) {
      body.source = payload.source;
    }
    if (payload.slug) {
      body.slug = payload.slug;
    }
    return await apiClient.post("/sites/", body);
  },
  republish: async (
    slug: string,
    payload?: Pick<PublishSitePayload, "source" | "visibility">
  ): Promise<PublishedSite> => {
    const body: Record<string, unknown> = {};
    if (payload?.source) {
      body.source = payload.source;
    }
    if (payload?.visibility) {
      body.visibility = payload.visibility;
    }
    return await apiClient.post(`/sites/${encodeURIComponent(slug)}/republish`, body);
  },
  unpublish: async (slug: string): Promise<PublishedSite> => {
    return await apiClient.delete(`/sites/${encodeURIComponent(slug)}`);
  },
};

File: backend/app/schemas/sites.py
"""Published-site API request/response models."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

DEFAULT_SITE_SOURCE = "/mnt/user-data/outputs/app-dist"
SiteVisibility = Literal["unlisted", "public"]


class PublishSiteRequest(BaseModel):
    """Create or republish a site for a conversation."""

    conversation_id: str = Field(min_length=1)
    source: str = Field(default=DEFAULT_SITE_SOURCE, min_length=1)
    slug: str | None = Field(
        default=None,
        description="仅前端自定义链接时传入；省略则服务端生成",
    )
    visibility: SiteVisibility = "unlisted"


class RepublishSiteRequest(BaseModel):
    """Re-snapshot an existing slug."""

    source: str = Field(default=DEFAULT_SITE_SOURCE, min_length=1)
    visibility: SiteVisibility | None = None


class PublishedSiteData(BaseModel):
    """Public representation of a published site."""

    slug: str
    version: int
    url: str
    visibility: str
    size_bytes: int
    expires_at: datetime | None
    conversation_id: str
    unpublished_at: datetime | None = None
    entry: str = "index.html"
    file_count: int | None = None
